[Build] build for catalyst support will generate invalid compile command

### Describe the issue

building for catalyst will fail on compiling onnruntime_mlas

`[ 44%] Building CXX object CMakeFiles/onnxruntime_mlas.dir/.../onnxruntime/onnxruntime/core/mlas/lib/platform.cpp.o
c++: error: version ' -ffunction-sections -fdata-sections' in target triple 'arm64-apple-ios13.3-macabi -ffunction-sections -fdata-sections' is invalid
c++: warning: overriding '-mmacosx-version-min=13.3' option with '--target=arm64-apple-ios13.3-macabi -ffunction-sections -fdata-sections' [-Woverriding-option]`

this was tested with `main` on `5746ba9d3b7b5eaf3a5c64fd24974f3649d71b34`




### Urgency

_No response_

### Target platform

macos 15.4.1

### Build script

`python3 onnxruntime/tools/ci_build/build.py --update --build --build_dir build-onnxruntime-ios-catalyst-Release-3.28.6-nosign --config Release --cmake_path cmake-3.28.6-macos-universal/CMake.app/Contents/bin/cmake --parallel 10 --compile_no_warning_as_error --skip_submodule_sync --disable_ml_ops --disable_rtti --minimal_build --cmake_extra_defines CMAKE_IGNORE_PREFIX_PATH=/opt/homebrew --osx_arch arm64 --build_apple_framework --macos Catalyst --apple_sysroot macosx --apple_deploy_target 13.3 --path_to_protoc_exe /opt/homebrew/opt/protobuf@21/bin/protoc`

### Error / output

`[ 44%] Building CXX object CMakeFiles/onnxruntime_mlas.dir/.../onnxruntime/onnxruntime/core/mlas/lib/platform.cpp.o
c++: error: version ' -ffunction-sections -fdata-sections' in target triple 'arm64-apple-ios13.3-macabi -ffunction-sections -fdata-sections' is invalid
c++: warning: overriding '-mmacosx-version-min=13.3' option with '--target=arm64-apple-ios13.3-macabi -ffunction-sections -fdata-sections' [-Woverriding-option]`


### Visual Studio Version

_No response_

### GCC / Compiler Version

Apple clang version 17.0.0 (clang-1700.0.13.5)

this seems to resolve the initial issue, but it still fails with:
`[ 79%] Linking CXX executable onnxruntime_mlas_q4dq
ld: building for 'macCatalyst', but linking in object file (/Users/nloriant/dev/myscript_ml/build-onnxruntime-ios-catalyst-Release-3.28.6-nosign/Release/libonnxruntime_mlas.a[35](ConvSymS8KernelDot.S.o)) built for 'macOS'
c++: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [onnxruntime_mlas_q4dq] Error 1
make[1]: *** [CMakeFiles/onnxruntime_mlas_q4dq.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....`

--- a/cmake/onnxruntime_mlas.cmake
+++ b/cmake/onnxruntime_mlas.cmake
@@ -790,12 +790,6 @@ if (WIN32)
   endif()
 endif()
 
-if (PLATFORM_NAME STREQUAL "macabi")
-  # Needed for maccatalyst C compilation
-  # i.e. the flags below add "--target=x86_64-apple-ios14.0-macabi -ffunction-sections -fdata-sections"
-  target_compile_options(onnxruntime_mlas PRIVATE ${CMAKE_C_FLAGS})
-endif()
-
 if (NOT onnxruntime_BUILD_SHARED_LIB)
     install(TARGETS onnxruntime_mlas EXPORT ${PROJECT_NAME}Targets
             ARCHIVE   DESTINATION ${CMAKE_INSTALL_LIBDIR}

--- a/tools/ci_build/build.py
+++ b/tools/ci_build/build.py
@@ -834,8 +834,6 @@ def generate_build_tree(
 
     if is_macOS() and not args.android:
         add_default_definition(cmake_extra_defines, "CMAKE_OSX_ARCHITECTURES", args.osx_arch)
-        if args.apple_deploy_target:
-            cmake_args += ["-DCMAKE_OSX_DEPLOYMENT_TARGET=" + args.apple_deploy_target]
         # Code sign the binaries, if the code signing development identity and/or team id are provided
         if args.xcode_code_signing_identity:
             cmake_args += ["-DCMAKE_XCODE_ATTRIBUTE_CODE_SIGN_IDENTITY=" + args.xcode_code_signing_identity]
@@ -926,7 +924,6 @@ def generate_build_tree(
         cmake_args += [
             "-Donnxruntime_BUILD_SHARED_LIB=ON",
             "-DCMAKE_OSX_SYSROOT=" + args.apple_sysroot,
-            "-DCMAKE_OSX_DEPLOYMENT_TARGET=" + args.apple_deploy_target,
             # we do not need protoc binary for ios cross build
             "-Dprotobuf_BUILD_PROTOC_BINARIES=OFF",
             "-DPLATFORM_NAME=" + platform_name,
@@ -942,16 +939,14 @@ def generate_build_tree(
         if args.macos == "Catalyst":
             macabi_target = f"{args.osx_arch}-apple-ios{args.apple_deploy_target}-macabi"
             cmake_args += [
-                "-DCMAKE_CXX_COMPILER_TARGET=" + macabi_target,
-                "-DCMAKE_C_COMPILER_TARGET=" + macabi_target,
-                "-DCMAKE_CC_COMPILER_TARGET=" + macabi_target,
                 f"-DCMAKE_CXX_FLAGS=--target={macabi_target}",
-                f"-DCMAKE_CXX_FLAGS_RELEASE=-O3 -DNDEBUG --target={macabi_target}",
                 f"-DCMAKE_C_FLAGS=--target={macabi_target}",
-                f"-DCMAKE_C_FLAGS_RELEASE=-O3 -DNDEBUG --target={macabi_target}",
-                f"-DCMAKE_CC_FLAGS=--target={macabi_target}",
-                f"-DCMAKE_CC_FLAGS_RELEASE=-O3 -DNDEBUG --target={macabi_target}",
             ]
+        else:
+            cmake_args += [
+                "-DCMAKE_OSX_DEPLOYMENT_TARGET=" + args.apple_deploy_target,
+            ]
+
         if args.visionos:
             cmake_args += [
                 "-DCMAKE_SYSTEM_NAME=visionOS",

--- a/tools/ci_build/build.py
+++ b/tools/ci_build/build.py
@@ -941,6 +941,7 @@ def generate_build_tree(
             cmake_args += [
                 f"-DCMAKE_CXX_FLAGS=--target={macabi_target}",
                 f"-DCMAKE_C_FLAGS=--target={macabi_target}",
+                f"-DCMAKE_ASM_FLAGS=--target={macabi_target}",
             ]
         else:
             cmake_args += [

